Add a URI decoder, with tests

levitte opened the pull request:
<!--
Thank you for your pull request. Please review below requirements.

Contributors guide: https://github.com/openssl/openssl/blob/master/CONTRIBUTING
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [ ] documentation is added or updated
- [x] tests are added or updated
- [x] CLA is signed

##### Description of change
<!-- Provide a description of the changes.

If it fixes a github issue, add Fixes #XXXX.
-->

The test uses `test/uritests.txt`.  That file could use some more tests, please suggest some.

Fixes #1957

levitte committed:
Add a URI decoder

This is a needed component for a URI based STORE module.

This includes a perl script crypto/uri_charmap.pl that generates a
local header file to codify the URI character classes.  See RFC3986
for further reference.

levitte committed:
Add a URI decoder internals test

This becomes part of the tests run by test/recipes/03-test_internal.t

richsalz commented:
have you looked at any tests from w3c?  Google it, there's a bunch.

levitte commented:
Most tests are about uri joining, which this module doesn't do. 
I finally found a file with a large number of "interesting"  uris, though. I'm currently working through it to fill uritests.txt

levitte commented:
Ok, that's a load of test URIs, and the bugs they helped clear out fixed

levitte commented:
Rebased on top of a fresher master

levitte commented:
Update: there is now a function to decode the authority into its smaller parts, *user*, *password*, *host* and *service*, as well as a function to do percent decoding inline.

mattcaswell commented on crypto/uri.c:
```diff
@@ -0,0 +1,836 @@
+                 * maxcolumns allowing, and we might also end this with a last
+                 * serie of HEXDIG.  We try IPv4 first.
+                 */
+                else {
```
else has { so if must have one too

levitte replied:
Okie

levitte replied:
I refuse to indent the comment above it, though.  That's just odd in this flow

mattcaswell commented on crypto/uri.c:
```diff
@@ -0,0 +1,836 @@
+
+    *path = NULL;
+    while(span_slash_segment(p, &off))
+        ;
```
continue

mattcaswell commented on crypto/uri.c:
```diff
@@ -0,0 +1,836 @@
+        off++;
+        if (span_segment_nz(p, &off)) {
+            while(span_slash_segment(p, &off))
+                ;
```
continue

mattcaswell commented on crypto/uri.c:
```diff
@@ -0,0 +1,836 @@
+    p++;
+
+    while(span_pchars(p, &off) || span_chars(p, &off, 0, "/?"))
+        ;
```
continue

mattcaswell commented on crypto/uri.c:
```diff
@@ -0,0 +1,836 @@
+    p++;
+
+    while(span_pchars(p, &off) || span_chars(p, &off, 0, "/?"))
+        ;
```
continue

levitte commented on test/uritests.txt:
```diff
@@ -0,0 +1,2211 @@
+######################################################################
+#
+# Shamelessly stolen from
+# https://github.com/cweb/iri-tests/blob/master/iris.txt
```
[I asked](https://github.com/cweb/iri-tests/issues/1)

cweb replied:
It's an MIT license, I hope you find it useful.  You may want to use urls.json instead though, if you can deal with this format:  https://github.com/cweb/url-testing/blob/master/urls.json  It's an updated version of iris.json, which iris.txt derives from.

levitte commented on crypto/uri.c:
```diff
@@ -0,0 +1,836 @@
+               ? c - 'A' + 10
+               : (c >= 'a' && c <= 'f'
+                  ? c - 'a' + 10
+                  : -1)));
```
And as I just discovered, an unnecessary function (we have `OPENSSL_hexchar2int` already)

levitte replied:
Agh!  No!  `OPENSSL_hexchar2int()` does a toebcdic conversion, which is *way wrong* in this case.

levitte replied:
I made `dehex` less hellish, though.

levitte commented on test/uri_internal_test.c:
```diff
@@ -0,0 +1,522 @@
+        linecounter++;
+
+        for (p = line; *p != '\n' && isspace(*p); p++)
+            ;
```
okie

levitte commented on test/uri_internal_test.c:
```diff
@@ -0,0 +1,522 @@
+
+        if (*p == '\n') {
+            if (test_data != NULL)
+                break;
```
`read_test_data()` reads one stanza per call.  Stanzas are delimited by empty lines.  So, if it finds an empty line and `test_data` is non-NULL, it means it's finished reading one stanza and must break out of the loop.

richsalz replied:
add skip lines starting with ```#```

levitte commented on test/uri_internal_test.c:
```diff
@@ -0,0 +1,522 @@
+        if (skip_stanza)
+            continue;
+
+        if (strcasecmp(key, "uri") == 0)
```
Yeahok

richsalz commented on crypto/uri_charmap.h:
```diff
@@ -0,0 +1,47 @@
+static const unsigned short char_type[] = {
+       0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
+       0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
+       0,    0,    0,    0,    0,    0,    0,    0,    0,   32,    0,   16,
```
it would be more clear if these values were in hex.

richsalz commented on crypto/uri_charmap.h:
```diff
@@ -0,0 +1,47 @@
+ * https://www.openssl.org/source/license.html
+ */
+
+#define CHARTYPE_ALPHA          1
```
max these hex.

richsalz commented on test/uri_internal_test.c:
```diff
@@ -0,0 +1,522 @@
+}
+
+static int linecounter = 0;
+static TEST_DATA *read_test_data(FILE *f)
```
this is a useful function; e.g., bntest and exptest could use it.

levitte commented on test/uri_internal_test.c:
```diff
@@ -0,0 +1,522 @@
+                OPENSSL_free(value);
+            }
+        else if (strcasecmp(key, "scheme") == 0
+                 || strcasecmp(key, "decoded_scheme") == 0)
```
A macro, methinks.  They touch a number of local variables

levitte commented on test/uri_internal_test.c:
```diff
@@ -0,0 +1,522 @@
+            || OPENSSL_percent_decode_inline(service));
+    if (!rv) {
+        ERR_print_errors_fp(stderr);
+        goto err;
```
Good point.

richsalz reviewed (commented):
you find it useful/interesting to look at the stanza stuff in bntest.c

richsalz commented:
Actually, perhaps go the other say.  Look at test/bntest.c, which has a generic key/value reader.  Pull that out and then use it in your code here, with semantics on top of the PAIRs.

levitte commented on test/uri_internal_test.c:
```diff
@@ -0,0 +1,522 @@
+    TEST_DATA *test_data = NULL;
+    int errs = 0;
+
+    OPENSSL_init_crypto(OPENSSL_INIT_LOAD_CRYPTO_STRINGS, NULL);
```
'cause they aren't loaded by default.  But maybe that should be done in `crypto/uri.c`, instead of `OPENSSL_init_crypto(0, NULL);`?

mattcaswell replied:
Err, yeah they are (that's kind of the point of auto-init):

https://github.com/openssl/openssl/blob/master/crypto/err/err.c#L663

levitte replied:
Okie...  I don't remember, frankly, except that I *did* run into trouble at some early stage and this line fixed it.  I'll try and see what happens when I remove it.

levitte replied:
Experiment done, and works.  This line is byebye

levitte commented on crypto/uri.c:
```diff
@@ -0,0 +1,836 @@
+    if (!is_ALPHA(p[off]))
+        return 1;
+    while (span_chars(p, &off, CHARTYPE_ALPHA|CHARTYPE_DIGIT, "+-."))
+        ;
```
Okie.  I found all other similar ones and replaced added `continue` on them as well.

levitte commented on crypto/uri.c:
```diff
@@ -0,0 +1,836 @@
+    OPENSSL_init_crypto(0, NULL);
+
+    if (extract_scheme(scheme, &p)
+        && (*scheme
```
Okie

levitte committed:
[to be squashed] fixups of crypto/uri.c

levitte committed:
[to be squashed] fixups of test/uri_internal_test.c

levitte commented:
Ok, @mattcaswell, I think I've responded to all your comments.

levitte commented:
The better idea would be to make a stanza library for everyone to use instead of doing reuse-by-copy.

richsalz commented:
well, yeah, that's what i was suggesting you do :)

levitte commented:
Ah.  Frankly, I'd rather do that refactor in a separate PR.

levitte commented:
@richsalz, I noticed your comments about `uri_charmap`, even though github seems to have lost them.

levitte committed:
[to be squashed] fixups of test/uri_internal_test.c

levitte committed:
[to be squashed] fixups of crypto/uri_charmap.*

levitte committed:
[to be squashed] fixups of test/uri_internal_test.c

mattcaswell commented on crypto/uri.c:
```diff
@@ -0,0 +1,842 @@
+                            if (++dblcolon > 1)
+                                break;
+                            maxcolons -= 2;
+                        } else
```
{} on else

levitte replied:
Yup

mattcaswell commented on crypto/uri.c:
```diff
@@ -0,0 +1,842 @@
+
+    /* Check if the loop stopped because of an invalid character */
+    if (!is_valid(p[off])) {
+        char tmpbuf[200];
```
newline after this

levitte replied:
Okie

mattcaswell commented on crypto/uri.c:
```diff
@@ -0,0 +1,842 @@
+{
+    const char *p = uri;
+
+    OPENSSL_init_crypto(0, NULL);
```
I'm wondering if this is actually necessary?

levitte replied:
It's been some time, I do recall there was a reason but cannot remember what it was...

mattcaswell replied:
It would be good to know. The thing is we are supposed to auto-initialise if we call any functions that need it. These functions don't seem to need it themselves (e.g. no shared global data etc), so if this call was necessary then it suggests some other function call was made that didn't auto-initialise when (possibly) it should have done.

levitte replied:
It might be something in the STORE effort that got jinxed...  Anyway, I've just taken those inits away, things are still working fine, so...

mattcaswell commented on crypto/uri.c:
```diff
@@ -0,0 +1,842 @@
+{
+    const char *p = authority;
+
+    OPENSSL_init_crypto(0, NULL);
```
and here

mattcaswell commented on test/uri_internal_test.c:
```diff
@@ -0,0 +1,528 @@
+        } else if (strcasecmp(key, "scheme") == 0
+                 || strcasecmp(key, "decoded_scheme") == 0) {
+            if (test_data != NULL && test_data->expected.scheme == NULL) {
+                test_data->expected.scheme = value;
```
Weren't you going to do this stuff in a macro?

levitte replied:
Oh, right!

mattcaswell reviewed (changes requested):
I took another scan through this and picked up a few more nits.

The main stumbling block for me on this is the licence question.

levitte committed:
[to be squashed] fixups of crypto/uri.c

levitte committed:
[to be squashed] fixups of test/uri_internal_test.c

levitte commented:
As for licensing, if we don't get an answer, I'll simply take it all away and we'll just add new cases as time goes.

levitte committed:
[to be squashed] fixups of crypto/uri.c, take away OPENSSL_init_crypto calls

levitte commented:
I just discovered that the IPv6 address parser failed on some perfectly acceptable addresses...

levitte committed:
[to be squashed] fixups of crypto/uri.c, fix IPv6 parsing

levitte committed:
[to be squashed] New set of test cases, invented in house or taken from a document under CC0

levitte commented:
I rewrote `test/uritests.txt` to only include URIs I've written myself, or some taken from a CC0 licensed document (verifiable)

Also fixed a few errors in the IPv6address check

levitte commented:
With a simplification of #2011, this turned out not to be necessary any more.

@cweb, sorry about the noise.

dwmw2 commented:
This is abandoned, right? If it comes back, does it need to support RFC7512 URIs too? 

I was actually inclined to just use libp11-kit for a PKCS#11 STORE so I don't really need it, but if OpenSSL *has* a URI parser then it would make sense for me to use it.

levitte commented:
It is abandoned, yes.
